File: DAA/DAA cpp/DAA Pract4.cpp
#include <iostream>
#include <vector>

using namespace std;

// Function to solve 0-1 Knapsack problem using dynamic programming
int knapsackDP(int capacity, const vector<int>& weights, const vector<int>& values, int n) {
    vector<vector<int>> dp(n + 1, vector<int>(capacity + 1, 0));

    for (int i = 1; i <= n; ++i) {
        for (int w = 1; w <= capacity; ++w) {
            if (weights[i - 1] <= w) {
                dp[i][w] = max(values[i - 1] + dp[i - 1][w - weights[i - 1]], dp[i - 1][w]);
            } else {
                dp[i][w] = dp[i - 1][w];
            }
        }
    }

    return dp[n][capacity];
}

int main() {
    int n;
    cout << "Enter the number of items: ";
    cin >> n;

    vector<int> weights(n);
    vector<int> values(n);

    cout << "Enter the weights of items:\n";
    for (int i = 0; i < n; ++i) {
        cout << "Item " << i + 1 << ": ";
        cin >> weights[i];
    }

    cout << "Enter the values of items:\n";
    for (int i = 0; i < n; ++i) {
        cout << "Item " << i + 1 << ": ";
        cin >> values[i];
    }

    int capacity;
    cout << "Enter the knapsack capacity: ";
    cin >> capacity;

    int maxValue = knapsackDP(capacity, weights, values, n);
    cout << "Maximum value that can be obtained: " << maxValue << endl;

    return 0;
}

File: DAA/DAA python/DAA Pract4.py
def knapsack_dp(weights , values ,capacity):
    n=len(values)
    
    dp = [[0 for _ in range(capacity + 1)] for _ in range(n + 1)]

    for i in range(1 , n+1):
        for w in range(1, capacity+1):
            if weights[i-1] <=w:
                dp[i][w] = max(values[i-1] + dp[i-1] [w - weights[i-1]], dp[i-1][w])

            else: 
                dp[i][w] = dp[i-1][w]

    return dp[n][capacity]

n= int(input("Enter the number of items: "))
weights = [int(input(f"Enter the weight of item({i+1}: ")) for i in range(n)]
values = [int(input(f"Enter the value of item {i+1}: ")) for i in range(n)]
capacity = int(input("Enter the knapsack capacity: "))

result = knapsack_dp(weights , values , capacity)

print(f"The maximum value that can be obtained is: {result}")